support decimal scalar value

liukun4515 opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #1393 

From #122 In order to support the decimal data type in the datafusion, we should add decimal scalar value first.

 # Rationale for this change
<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

# What changes are included in this PR?
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

# Are there any user-facing changes?
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

liukun4515 commented:
```#[allow(clippy::box_vec)]```
I am confused about the failed [ci](https://github.com/apache/arrow-datafusion/runs/4403890233?check_suite_focus=true)

liukun4515 commented:
@Dandandan @alamb @houqp PTAL

alamb commented:
> I am confused about the failed ci

 this is likely due to a new rust version being released with more stringent clippy lint rules

Thankfully @xudong963  has a PR up to get it working again: https://github.com/apache/arrow-datafusion/pull/1395

alamb commented on datafusion/src/scalar.rs:
```diff
@@ -43,6 +43,8 @@ pub enum ScalarValue {
     /// 64bit float
     Float64(Option<f64>),
+    /// 128bit decimal, using the i128 to represent the decimal
+    Decimal128(Option<i128>, Option<usize>, Option<usize>),
```
🤔  it seems like we always need the precision and scale (to know what type the Decimal is). 

Thus, perhaps this could be changed to be

```suggestion
    Decimal128(Option<i128>, usize, usize),
```

liukun4515 replied:
nice point.
I almost forgot this.

alamb commented on datafusion/src/scalar.rs:
```diff
@@ -453,6 +477,26 @@ macro_rules! eq_array_primitive {
+        scale: usize,
+    ) -> Result<Self> {
+        // make sure the precision and scale is valid
+        // TODO const the max precision and min scale
```
The arrow spec doesn't seem to say anything about min/max valid scales: https://github.com/apache/arrow/blob/master/format/Schema.fbs#L181-L190

I poked a little around in arrow-rs and I also didn't find any limits on the precision or scale either

liukun4515 replied:
If we use 128bit to represent a decimal value, the max `precision` is 38.
The `scale` is always greater or equal to 0 and less than or equal to `precision`.
In the arrow-rs, the decimal is represented by `i128` with `precision` and `scale`.
So we can add the `MAX_PRECISION = 38` in the datafusion.

capkurmagati replied:
Just to make sure I got this correctly.
The spec doesn't say anything about the the precision/scale of `DECIMAL`, but we're implementing `DECIMAL128` here so 38 should be the max precision.
After https://github.com/apache/arrow-rs/issues/131 is implemented, we can implement `DECIMAL256` type in datafusion and the max precision will be 76.

liukun4515 replied:
> 

Yes, now in the arrow-rs the decimal256 has not been implemented.
In the arrow-go, it is also not implemented too.
Just in the arrow-java and arrow-c++, the bitwith of 256 has been implemented.
@capkurmagati

liukun4515 replied:
we can make this const in other pull request

alamb commented on datafusion/src/scalar.rs:
```diff
@@ -100,6 +102,12 @@ impl PartialEq for ScalarValue {
         match (self, other) {
+            (Decimal128(v1, p1, s1), Decimal128(v2, p2, s2)) => {
+                v1.eq(v2) && p1.eq(p2) && s1.eq(s2)
+                // TODO how to handle this case: decimal(123,10,1) with decimal(1230,10,2)
```
`Int8(Some(100))` is not equal to `Int64(Some(100))` with this implementation either , so I think it is consistent with the rest of this comparison of `decimal(123,10,1) and `decimal(1230,10,2)` are different --

liukun4515 replied:
yep.
If we want to compare two values with the different data types, you should convert them to the same data type and then compare them.

alamb commented on datafusion/src/scalar.rs:
```diff
@@ -171,6 +179,17 @@ impl PartialOrd for ScalarValue {
         // any newly added enum variant will require editing this list
         // or else face a compile error
         match (self, other) {
+            // TODO decimal type, we just compare the values which have the same precision and scale.
```
I think the code in this PR is correct  -- two decimal's can be compared if they have the same precision / scale and they can't be compared (returns `None`) if they have different precision/scales).

alamb commented on datafusion/src/scalar.rs:
```diff
@@ -253,6 +272,11 @@ impl std::hash::Hash for ScalarValue {
         match self {
+            Decimal128(v, p, s) => {
+                v.hash(state);
+                p.hash(state);
+                s.hash(state)
```
I think we could probably skip hashing the precision and scale - and just hash the value

liukun4515 replied:
If two decimal values have the same value and diff precision or scale, the hash value is the same.
But if we hash all data in the decimal, the above case will not happen.
It is better to hash all data in decimal.

alamb replied:
That is a good point 👍

alamb commented on datafusion/src/scalar.rs:
```diff
@@ -465,6 +509,14 @@ impl ScalarValue {
+            ScalarValue::Decimal128(_, Some(precision), Some(scale)) => {
+                DataType::Decimal(*precision, *scale)
+            }
+            ScalarValue::Decimal128(_, _, _) => {
+                // TODO add the default precision and scale for this case
+                // DataType::Decimal(38, 0)
+                panic!("The Decimal Scalar value with invalid precision or scale.");
```
I think if you changed Decimal as suggested above to always have precision and scale, this code would not be possible

alamb commented on datafusion/src/scalar.rs:
```diff
@@ -831,13 +892,40 @@ impl ScalarValue {
+                ScalarValue::Decimal128(v1, _, _) => v1,
+                _ => unreachable!(),
+            })
+            .collect::<Vec<Option<i128>>>();
```
Using `DecimalBuilder` in this way is awkward. I wonder if we could add a function to create Decimal values from an array of `i128` (likely it would go in arrow-rs). Something like

```
let arr = DecimalArray::from_iter_and_scale(array.into_iter(), precision, scale)
```

liukun4515 replied:
> 

Yes, we can add the `from_iter_and_scale` function in arrow-rs later and replace this in the followup pull request.
Do you agree this? @alamb

liukun4515 replied:
related pr: https://github.com/apache/arrow-rs/issues/1009

alamb commented on datafusion/src/scalar.rs:
```diff
@@ -1061,19 +1176,41 @@ impl ScalarValue {
+        scale: &usize,
+    ) -> ScalarValue {
+        let array = array.as_any().downcast_ref::<DecimalArray>().unwrap();
+        // TODO add checker: the precision and scale are same with array
```
this would be a good `assert!` type check to add, though I don't think it would ever happen unless there is some bug

alamb reviewed (commented):
This is looking good @liukun4515  -- thank you. I had a few suggestions, but overall this looks just about perfect.

liukun4515 commented:
@alamb  I have addressed all comments.
If there are no other comments, please help to merge this pr.
I will propose other pull requests about decimal, for example aggregate with decimal data type.

alamb commented on datafusion/src/scalar.rs:
```diff
@@ -43,6 +43,8 @@ pub enum ScalarValue {
     /// 64bit float
     Float64(Option<f64>),
+    /// 128bit decimal, using the i128 to represent the decimal
+    Decimal128(Option<i128>, usize, usize),
```
👍

alamb reviewed (approved):
This looks great -- thank you @liukun4515
